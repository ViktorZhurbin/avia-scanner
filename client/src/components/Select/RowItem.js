import React from 'react';
import PropTypes from 'prop-types';
import cl from 'classnames/bind';

import { codeNamePropType } from '../../entities/propTypes';
import styles from './index.css';

const cx = cl.bind(styles);

class RowItem extends React.Component {
    static propTypes = {
        value: codeNamePropType.isRequired,
        index: PropTypes.number.isRequired,
        onSelect: PropTypes.func.isRequired,
        children: PropTypes.node.isRequired,
        isSelected: PropTypes.bool,
        isDisabled: PropTypes.bool,
    }

    static defaultProps = {
        isSelected: false,
        isDisabled: false,
    }

    onSelect = () => {
        const { value, onSelect } = this.props;

        onSelect(value);
    }

    onKeyPress = (event) => {
        if (event.key === 'Enter') {
            this.onSelect();
        }
    }

    render() {
        const {
            value,
            index,
            children,
            isSelected,
            isDisabled,
        } = this.props;

        return (
            <div
                key={value.code}
                className={cx('item', {
                    isSelected,
                    isDisabled,
                })}
                role="button"
                tabIndex={index + 1}
                onKeyPress={this.onKeyPress}
                onClick={this.onSelect}
            >
                {children}
            </div>
        );
    }
}

export default RowItem;
