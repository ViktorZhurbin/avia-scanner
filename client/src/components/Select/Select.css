.itemList.currency {
    position: absolute;
    top: 100%;
    right: 0;
    margin-top: 10px;
    width: var(--width);
    box-shadow: 0 0 15px rgba(0, 0, 0, 0.25);
    padding: 15px 0;
    background: white;
    border-radius: 5px;
}

.itemList.formSelect {
    width: 100%;
    position: absolute;
    top: 68px;
    box-shadow: 0 0 15px rgba(0, 0, 0, 0.25);
    padding: 15px 0;
    background: white;
    border-radius: 0 0 5px 5px;
    overflow-y: auto;
}

@media (min-width: 640px) {
    .itemList.formSelect {
        width: calc(100% - 2px);
    }
}
