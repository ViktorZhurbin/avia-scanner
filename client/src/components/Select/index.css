.itemList.currency {
    position: absolute;
    top: 100%;
    right: 0;
    margin-top: 10px;
    width: var(--width);
    box-shadow: 0 0 15px rgba(0, 0, 0, 0.25);
    padding: 15px 0;
    background: white;
    border-radius: 5px;
}

.itemList.formSelect {
    width: 100%;
    position: absolute;
    top: 68px;
    box-shadow: 0 0 15px rgba(0, 0, 0, 0.25);
    padding: 15px 0;
    background: white;
    border-radius: 0 0 5px 5px;
    overflow-y: auto;
}

.item {
    color: var(--color-twilight);

    &.isSelected {
        color: var(--color-blue);
    }

    &.isDisabled {
        color: var(--color-metal);
        pointer-events: none;
    }
}

@media (min-width: 640px) {
    .itemList.formSelect {
        width: calc(100% - 2px);
    }
}
