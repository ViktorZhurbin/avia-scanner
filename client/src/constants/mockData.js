export const places = [
    {
        code: 'TXL',
        name: 'Berlin',
    },
    {
        code: 'DME',
        name: 'Moscow',
    },
    {
        code: 'DUB',
        name: 'Dublin',
    },
    {
        code: 'HEL',
        name: 'Helsinki',
    },
    {
        code: 'TBS',
        name: 'Tbilisi',
    },
];

export const currencyList = [
    {
        code: 'USD',
        name: 'US Dollar',
    },
    {
        code: 'RUB',
        name: 'Rouble',
    },
    {
        code: 'EUR',
        name: 'Euro',
    },
];

export const currencyRates = (baseCurrency) => {
    const map = {
        USD: {
            USD: 1,
            EUR: 0.8,
            RUB: 60,
        },
        EUR: {
            USD: 1.1,
            EUR: 1,
            RUB: 70,
        },
        RUB: {
            USD: 0.016666667,
            EUR: 0.013,
            RUB: 1,
        },
    };

    return map[baseCurrency];
};
