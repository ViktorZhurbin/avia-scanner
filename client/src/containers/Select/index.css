.triggerContainer {
    box-sizing: border-box;
    background-color: var(--color-white);
    margin-bottom: 2px;
    margin-right: 0;
    cursor: pointer;
    height: 60px;

    &.isFirst {
        border-radius: 5px 5px 0 0;
    }
}

@media (min-width: 640px) {
    .triggerContainer {
        margin-right: 2px;
        margin-bottom: 0;

        &.isFirst {
            border-radius: 5px 0 0 5px;
        }
    }
}
