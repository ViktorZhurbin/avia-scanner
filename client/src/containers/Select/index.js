import React from 'react';
import PropTypes from 'prop-types';
import cl from 'classnames/bind';

import Dropdown from '../../components/Dropdown';
import SelectTrigger from '../../components/SelectTrigger';
import Place from './types/Place';
import Date from './types/Date';

import { formatDateByBrowserLocale } from '../../utils/string';

import styles from './index.css';

const cx = cl.bind(styles);

class Select extends React.Component {
    static propTypes = {
        type: PropTypes.string.isRequired,
        id: PropTypes.string,
        value: PropTypes.string,
        onSelect: PropTypes.func.isRequired,
        placeholder: PropTypes.string.isRequired,
        isFirst: PropTypes.bool,
        isLast: PropTypes.bool,
    }

    static defaultProps = {
        isFirst: false,
        isLast: false,
        value: null,
        id: null,
    }

    renderTrigger = () => {
        const {
            value,
            placeholder,
            isFirst,
            isLast,
            type,
        } = this.props;

        const formatted = (value && type === 'date')
            ? formatDateByBrowserLocale(value)
            : value;

        return (
            <div
                className={cx({
                    triggerContainer: true,
                    isFirst,
                    isLast,
                })}
            >
                <SelectTrigger
                    selectedItem={formatted}
                    placeholder={placeholder}
                    classNames={{
                        type,
                    }}
                />
            </div>
        );
    }

    render() {
        const { type } = this.props;

        const Component = type === 'place' ? Place : Date;

        return (
            <Dropdown
                classNames={{
                    placeSelect: true,
                }}
                trigger={this.renderTrigger()}
            >
                <Component {...this.props} />
            </Dropdown>
        );
    }
}

export default Select;
