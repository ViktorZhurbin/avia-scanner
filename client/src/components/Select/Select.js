import React from 'react';
import PropTypes from 'prop-types';
import cl from 'classnames/bind';

import Dropdown from '../Dropdown';
import RowItem from './RowItem';
import { classNamesPropType, codeNamePropType } from '../../entities/propTypes';

import styles from './Select.css';

const cx = cl.bind(styles);

const Select = (props) => {
    const {
        trigger,
        classNames,
        selectedItem,
        disabledItem,
        onSelect,
        itemList,
        renderItem,
    } = props;

    return (
        <Dropdown
            classNames={{ ...classNames }}
            trigger={trigger}
        >
            <div
                className={cx({
                    itemList: true,
                    ...classNames,
                })}
            >
                {itemList.map((item, index) => {
                    const isDisabled = (item && item.code) === (disabledItem && disabledItem.code);
                    const isSelected = (item && item.code) === (selectedItem && selectedItem.code);

                    return (
                        <RowItem
                            key={item.code}
                            className={classNames}
                            value={item}
                            index={index}
                            handleSelect={onSelect}
                            isSelected={isSelected}
                            isDisabled={isDisabled}
                        >
                            {renderItem(item)}
                        </RowItem>
                    );
                })}
            </div>
        </Dropdown>
    );
};

Select.propTypes = {
    trigger: PropTypes.node.isRequired,
    itemList: PropTypes.arrayOf(codeNamePropType).isRequired,
    renderItem: PropTypes.func.isRequired,
    selectedItem: codeNamePropType,
    disabledItem: codeNamePropType,
    onSelect: PropTypes.func.isRequired,
    classNames: classNamesPropType,
};

Select.defaultProps = {
    classNames: null,
    disabledItem: null,
    selectedItem: null,
};

export default Select;
