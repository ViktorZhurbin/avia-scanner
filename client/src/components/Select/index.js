import React from 'react';
import PropTypes from 'prop-types';
import cl from 'classnames/bind';

import Dropdown from '../Dropdown';
import RowItem from './RowItem';
import { classNamesPropType, codeNamePropType } from '../../entities/propTypes';
import styles from './index.css';

const cx = cl.bind(styles);

class Select extends React.PureComponent {
    static propTypes = {
        trigger: PropTypes.node.isRequired,
        itemList: PropTypes.arrayOf(codeNamePropType).isRequired,
        renderItem: PropTypes.func.isRequired,
        selectedItem: codeNamePropType,
        disabledItem: codeNamePropType,
        onSelect: PropTypes.func.isRequired,
        classNames: classNamesPropType,
    }

    static defaultProps = {
        classNames: null,
        disabledItem: null,
        selectedItem: null,
    }

    renderDropdown = () => {
        const {
            selectedItem,
            disabledItem,
            onSelect,
            itemList,
            renderItem,
            classNames,
        } = this.props;

        return (
            <div
                className={cx('itemList', ...classNames)}
            >
                {itemList.map((item, index) => {
                    const isDisabled = (item && item.code) === (disabledItem && disabledItem.code);
                    const isSelected = (item && item.code) === (selectedItem && selectedItem.code);

                    return (
                        <RowItem
                            key={item.code}
                            className={classNames}
                            value={item}
                            index={index}
                            onSelect={onSelect}
                            isSelected={isSelected}
                            isDisabled={isDisabled}
                        >
                            {renderItem(item)}
                        </RowItem>
                    );
                })}
            </div>
        );
    };

    render() {
        const { trigger, classNames } = this.props;
        return (
            <Dropdown
                classNames={classNames}
                trigger={trigger}
            >
                {this.renderDropdown()}
            </Dropdown>
        );
    }
}

export default Select;
