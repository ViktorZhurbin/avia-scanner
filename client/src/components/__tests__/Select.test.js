import React from 'react';
import { shallow } from 'enzyme';

import Select from '../Select';
import { places } from '../../constants/mockData';

const props = {
    itemList: places,
    selectedItem: places[0],
    disabledItem: places[1],
    onSelect: () => null,
};

describe('Render Select', () => {
    it('render correctly', () => {
        const component = shallow(
            <Select
                {...props}
                trigger={<div />}
                renderItem={() => <div />}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
