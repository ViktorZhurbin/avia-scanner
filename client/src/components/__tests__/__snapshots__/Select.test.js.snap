// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Select render correctly 1`] = `
<Dropdown
  classNames={Object {}}
  trigger={<div />}
>
  <div
    className="itemList"
  >
    <RowItem
      className={null}
      handleSelect={[Function]}
      index={0}
      isDisabled={false}
      isSelected={true}
      key="TXL"
      value={
        Object {
          "code": "TXL",
          "name": "Berlin",
        }
      }
    >
      <div />
    </RowItem>
    <RowItem
      className={null}
      handleSelect={[Function]}
      index={1}
      isDisabled={true}
      isSelected={false}
      key="DME"
      value={
        Object {
          "code": "DME",
          "name": "Moscow",
        }
      }
    >
      <div />
    </RowItem>
    <RowItem
      className={null}
      handleSelect={[Function]}
      index={2}
      isDisabled={false}
      isSelected={false}
      key="DUB"
      value={
        Object {
          "code": "DUB",
          "name": "Dublin",
        }
      }
    >
      <div />
    </RowItem>
    <RowItem
      className={null}
      handleSelect={[Function]}
      index={3}
      isDisabled={false}
      isSelected={false}
      key="HEL"
      value={
        Object {
          "code": "HEL",
          "name": "Helsinki",
        }
      }
    >
      <div />
    </RowItem>
    <RowItem
      className={null}
      handleSelect={[Function]}
      index={4}
      isDisabled={false}
      isSelected={false}
      key="TBS"
      value={
        Object {
          "code": "TBS",
          "name": "Tbilisi",
        }
      }
    >
      <div />
    </RowItem>
  </div>
</Dropdown>
`;
